File: python/revo3/subscriptions.py
"""Read finite State, optional Touch, and Health pull subscriptions."""

import argparse
import asyncio

from bc_revo3_sdk import main_mod as sdk


async def run(args: argparse.Namespace) -> None:
    manager = sdk.Manager()
    hand = None
    state_sub = None
    touch_sub = None
    health_sub = None
    try:
        hand = await manager.connect_auto(port=args.port, slave_id=args.slave_id)
        state_sub = hand.state.subscribe(period=args.period)
        for _ in range(args.count):
            state = await asyncio.wait_for(state_sub.next(), timeout=args.timeout)
            print(
                f"State timestamp={state.timestamp.sec}.{state.timestamp.nsec:09d} "
                f"J0={state.positions_deg[0]:.2f} degree"
            )

        if hand.touch.layout is not None:
            touch_sub = hand.touch.subscribe(period=args.period)
            for _ in range(args.count):
                frame = await asyncio.wait_for(touch_sub.next(), timeout=args.timeout)
                module_count = len(frame.modules)
                signal_modules = sum(
                    1
                    for module in frame.modules
                    if module.force3d is not None
                    or module.torque2d is not None
                    or module.resultant_force_mn is not None
                )
                print(
                    f"Touch sequence={frame.sequence} "
                    f"modules={module_count} force_modules={signal_modules}"
                )

        health_sub = hand.health.subscribe(period=args.health_period)
        for _ in range(args.count):
            health = await asyncio.wait_for(health_sub.next(), timeout=args.timeout)
            print(
                f"Health safety={health.safety_state} "
                f"system={health.system_state} error={health.system_error_code} "
                f"faulted_motors={health.faulted_motor_count}"
            )

        statistics = hand.statistics
        print(
            "RuntimeStatistics: "
            f"state_reads={statistics.state_reads}, "
            f"touch_reads={statistics.touch_reads}, "
            f"failed_operations={statistics.failed_operations}"
        )
    finally:
        if health_sub is not None:
            health_sub.close()
        if touch_sub is not None:
            touch_sub.close()
        if state_sub is not None:
            state_sub.close()
        if hand is not None:
            await hand.close()
        await manager.close()


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument("--port")
    parser.add_argument("--slave-id", type=lambda value: int(value, 0))
    parser.add_argument("--period", type=float, default=0.02)
    parser.add_argument("--health-period", type=float, default=1.0)
    parser.add_argument("--count", type=int, default=3)
    parser.add_argument(
        "--timeout",
        type=float,
        default=5.0,
        help="timeout in seconds for each subscription read (default: 5.0)",
    )
    args = parser.parse_args()
    if args.period <= 0 or args.health_period <= 0:
        parser.error("subscription periods must be positive")
    if args.count <= 0:
        parser.error("count must be positive")
    if args.timeout <= 0:
        parser.error("timeout must be positive")
    return args


if __name__ == "__main__":
    asyncio.run(run(parse_args()))

File: c/demo/subscriptions.cpp
#include <revo3/revo3.hpp>

#include <chrono>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <limits>
#include <stdexcept>
#include <string>

namespace {

long parse_positive(const char *value, const char *option) {
  char *end = nullptr;
  const auto parsed = std::strtol(value, &end, 10);
  if (value[0] == '\0' || end == value || *end != '\0' || parsed <= 0) {
    throw std::invalid_argument(std::string(option) + " must be positive");
  }
  return parsed;
}

std::uint8_t parse_slave_id(const char *value) {
  char *end = nullptr;
  const auto parsed = std::strtoul(value, &end, 0);
  if (value[0] == '\0' || end == value || *end != '\0' || parsed > 255) {
    throw std::invalid_argument("--slave-id must fit in uint8");
  }
  return static_cast<std::uint8_t>(parsed);
}

const char *require_value(int argc, char **argv, int &index,
                          const char *option) {
  if (index + 1 >= argc) {
    throw std::invalid_argument(std::string(option) + " requires a value");
  }
  return argv[++index];
}

void print_usage(const char *program) {
  std::printf(
      "Usage: %s [--port NAME] [--slave-id ID] [--period-ms MS] "
      "[--health-period-ms MS] [--count N]\n",
      program);
}

}  // namespace

int main(int argc, char **argv) {
  try {
    revo3::DiscoveryOptions discovery;
    auto period = std::chrono::milliseconds(20);
    auto health_period = std::chrono::milliseconds(1000);
    int count = 3;

    for (int index = 1; index < argc; ++index) {
      if (std::strcmp(argv[index], "-h") == 0 ||
          std::strcmp(argv[index], "--help") == 0) {
        print_usage(argv[0]);
        return 0;
      }
      if (std::strcmp(argv[index], "--port") == 0) {
        discovery.port = require_value(argc, argv, index, "--port");
      } else if (std::strcmp(argv[index], "--slave-id") == 0) {
        discovery.slave_id = parse_slave_id(
            require_value(argc, argv, index, "--slave-id"));
      } else if (std::strcmp(argv[index], "--period-ms") == 0) {
        period = std::chrono::milliseconds(parse_positive(
            require_value(argc, argv, index, "--period-ms"), "--period-ms"));
      } else if (std::strcmp(argv[index], "--health-period-ms") == 0) {
        health_period = std::chrono::milliseconds(parse_positive(
            require_value(argc, argv, index, "--health-period-ms"),
            "--health-period-ms"));
      } else if (std::strcmp(argv[index], "--count") == 0) {
        const auto value = parse_positive(
            require_value(argc, argv, index, "--count"), "--count");
        if (value > std::numeric_limits<int>::max()) {
          throw std::invalid_argument("--count is too large");
        }
        count = static_cast<int>(value);
      } else {
        throw std::invalid_argument(std::string("unknown option: ") +
                                    argv[index]);
      }
    }

    revo3::Manager manager;
    auto hand = manager.connect_auto(discovery);

    auto state_subscription = hand.state().subscribe(period);
    for (int index = 0; index < count; ++index) {
      const auto state = state_subscription.next();
      std::printf("State timestamp=%lld.%09lld J0=%.2f degree\n",
                  static_cast<long long>(state.timestamp.sec),
                  static_cast<long long>(state.timestamp.nsec),
                  state.motors.positions_deg[0]);
    }
    state_subscription.close();

    try {
      static_cast<void>(hand.touch().layout());
      auto touch_subscription = hand.touch().subscribe(period);
      for (int index = 0; index < count; ++index) {
        const auto frame = touch_subscription.next();
        std::size_t force_modules = 0;
        for (const auto &module : frame.modules) {
          if (module.has_force3d || module.has_torque2d ||
              module.has_resultant_force) {
            ++force_modules;
          }
        }
        std::printf("Touch sequence=%llu modules=%zu force_modules=%zu\n",
                    static_cast<unsigned long long>(frame.sequence),
                    frame.modules.size(), force_modules);
      }
      touch_subscription.close();
    } catch (const revo3::SdkError &error) {
      if (error.code() != revo3::SdkErrorCode::UnsupportedCapability) {
        throw;
      }
      std::printf("Touch subscription skipped: %s\n", error.what());
    }

    auto health_subscription = hand.health().subscribe(health_period);
    for (int index = 0; index < count; ++index) {
      const auto health = health_subscription.next();
      std::printf(
          "Health safety=%u system=%u error=%u faulted_motors=%u "
          "voltage=%u V\n",
          health.safety_state, health.system_state, health.system_error_code,
          health.faulted_motor_count, health.voltage_v);
    }
    health_subscription.close();

    const auto statistics = hand.statistics();
    std::printf(
        "RuntimeStatistics: state_reads=%llu touch_reads=%llu "
        "failed_operations=%llu\n",
        static_cast<unsigned long long>(statistics.state_reads),
        static_cast<unsigned long long>(statistics.touch_reads),
        static_cast<unsigned long long>(statistics.failed_operations));
    return 0;
  } catch (const revo3::SdkError &error) {
    std::fprintf(stderr, "Revo3 error: %s\n", error.what());
    return 1;
  } catch (const std::exception &error) {
    std::fprintf(stderr, "Error: %s\n", error.what());
    print_usage(argv[0]);
    return 1;
  }
}
